packages/SwingSet/docs/garbage-collection.md

```diff
@@ -64,46 +135,241 @@ When tracking the reachability state of a JavaScript `Object`, we define five st
+
+## Other Object Types
+
+Vats can import three kinds of things: objects, promises, and device nodes. Most GC is about objects.
+
+Objects have a strict and immutable notion of "importer" vs "exporter". Each object is exported by exactly one vat, and imported by zero or more vats (and/or referenced by kernel tables, independently of any vat).
+
+In contrast, Promises have a single "decider", which is either the one vat which currently holds decision-making authority, or the kernel itself (in the case of result promises for queued messages). The decider can move around over time, as the promise is used in the result of messages passed from one vat to another.
+
+In addition, the GC behavior of JS `Promise` objects is non-trivial. For unresolved Promises, vats need to remain prepared to handle a resolution as long as any `.then` callback is scheduled, but we have no way to sense whether `.then` has been called or not. Our general approach is to remove promise identifiers from each vat's c-list once they've been resolved, but keep a copy of the resolution data in the kernel until the promise ID has been fully retired.
+
+As a result, we track `Promise` objects in the liveslots `slotToVal` and `valToSlot` tables for the sake of serialization, but not in `exportedRemotables` or the `droppedRegistry`. We use a separate table named `pendingPromises` to keep the `Promise` objects alive until they are resolved.
+
+Vats can also import device nodes, however there are very few of them in the system (only a handful per device). Like Promises, we use `slotToVal` and `valToSlot`, and a special `importedDevices` table, but not the others. Devices nodes are immortal: vats never announce their drop or retirement.
+
+The `dropImports`, `retireImports`, and `retireExports` syscalls/deliveries only mention object references, never promises or device nodes.
+
+## Vat GC Cycle
+
+Within a vat, liveslots runs each delivery in two phases. The first phase is for userspace: a message or promise resolution is delivered, which causes arbitrary userspace code to run. Eventually that userspace code comes to a halt: userspace can use Promises to span multiple turns, but it does not get `setImmediate` or any other form of IO, so eventually control will return to liveslots, and userspace has no way to regain agency until liveslots makes another delivery.
+
+The second phase then begins with a call to the platform's `gc()` primitive, and enough delay to allow any activated finalizer callbacks to run. The necessary calls were determined experimentally for each JS engine we use: the winning combination for both Node.js and XS appears to be two `setImmediate` calls, followed by `gc()`, followed by a third `setImmediate`.
```

Comment:
Do we have a test of this recipe that we can use to re-validate when a new version of V8 is released?

Reply:
Yup: `test-gc-and-finalize.js`.

https://github.com/Agoric/agoric-sdk/blob/7b6024dcfbbc1a09fc3241d1e3e5a690e7fd1b3c/packages/SwingSet/test/test-gc-and-finalize.js